core/shared/src/main/scala/zio/Scope.scala

```diff
@@ -209,12 +229,25 @@ object Scope {
    * Snoyman @snoyberg.
    * (https://github.com/snoyberg/conduit/blob/master/resourcet/Control/Monad/Trans/Resource/Internal.hs)
    */
-  private abstract class ReleaseMap extends Serializable {
-
-    /**
-     * An opaque identifier for a finalizer stored in the map.
-     */
-    type Key
+  private final class ReleaseMap extends AtomicReference[State](State.initial) with Serializable {
+    ref: AtomicReference[State] =>
+    import State.{Exited, Running}
+
+    private[this] def next(l: Long) =
+      if (l == 0L) throw new RuntimeException("ReleaseMap wrapped around")
+      else if (l == Long.MinValue) Long.MaxValue
+      else l - 1
+
+    private[this] def modify[B](f: State => (UIO[B], State))(implicit trace: Trace): UIO[B] =
+      ZIO.suspendSucceed {
+        var b = null.asInstanceOf[UIO[B]]
+        while (b eq null) {
```

Comment:
This is one case I prefer recursion, but I will leave it up to you.

Reply:
I don't mind either way, but looking at the byte code AFAICT the implementation using a while-loop is slightly more performant as there is 1 less method invocation

Reply:
Let's keep it as is then.